src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs

```diff
@@ -0,0 +1,360 @@
+        let versioning = versioning_guard.deref_mut();
+        // Validate parameters.
+        if !versioning.current_version.levels.contains_key(&group_1) {
+            return Err(Error::CompactionGroup(format!("invalid group {}", group_1)));
+        }
+
+        if !versioning.current_version.levels.contains_key(&group_2) {
+            return Err(Error::CompactionGroup(format!("invalid group {}", group_2)));
+        }
+
+        let state_table_info = versioning.current_version.state_table_info.clone();
+        let mut member_table_ids_1 = state_table_info
+            .compaction_group_member_table_ids(group_1)
+            .iter()
+            .cloned()
+            .collect_vec();
+
+        let mut member_table_ids_2 = state_table_info
+            .compaction_group_member_table_ids(group_2)
+            .iter()
+            .cloned()
+            .collect_vec();
+
+        debug_assert!(!member_table_ids_1.is_empty());
+        debug_assert!(!member_table_ids_2.is_empty());
+        assert!(member_table_ids_1.is_sorted());
+        assert!(member_table_ids_2.is_sorted());
+
+        // Make sure `member_table_ids_1` is smaller than `member_table_ids_2`
+        let (left_group_id, right_group_id) =
+            if member_table_ids_1.first().unwrap() < member_table_ids_2.first().unwrap() {
+                (group_1, group_2)
+            } else {
+                std::mem::swap(&mut member_table_ids_1, &mut member_table_ids_2);
+                (group_2, group_1)
+            };
+
+        // We can only merge two groups with non-overlapping member table ids
+        if member_table_ids_1.last().unwrap() >= member_table_ids_2.first().unwrap() {
+            return Err(Error::CompactionGroup(format!(
+                "invalid merge group_1 {} group_2 {}",
+                left_group_id, right_group_id
+            )));
+        }
+
+        let combined_member_table_ids = member_table_ids_1
+            .iter()
+            .chain(member_table_ids_2.iter())
+            .collect_vec();
+        assert!(combined_member_table_ids.is_sorted());
+
+        // check duplicated sst_id
+        let mut sst_id_set = HashSet::new();
+        for sst_id in versioning
+            .current_version
+            .get_sst_ids_by_group_id(left_group_id)
+            .chain(
+                versioning
+                    .current_version
+                    .get_sst_ids_by_group_id(right_group_id),
+            )
+        {
+            if !sst_id_set.insert(sst_id) {
+                return Err(Error::CompactionGroup(format!(
+                    "invalid merge group_1 {} group_2 {} duplicated sst_id {}",
+                    left_group_id, right_group_id, sst_id
+                )));
+            }
+        }
+
+        // TODO(li0k): remove this check (Since the current split_sst does not change key_range, this check can not be removed, otherwise concate will fail.)
+        // check branched sst on non-overlap level
```

Comment:
nits: this comment is stale?

Reply:
Indeed, we can always maintain this check